### PR TITLE
Suppress logging from AzureSdk and GraphApi in Pulumi dynamic components

### LINKED ISSUE #1971: Suppress messages from dynamic components
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a request for a new feature in the Data Safe Haven or an upgrade to an existing feature.
- [x] The feature is still missing in the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :strawberry: Suggested change

Log messages produced by dynamic components make it into the Pulumi log and make it unnecessarily messy

## :steam_locomotive: How could this be done?

Suppress logging in `AzureApi` (as was done in the past) or make this configurable.

### PR DESCRIPTION
**Pull request by jemrobinson**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :vertical_traffic_light: Depends on

n/a

### :arrow_heading_up: Summary

Redirects AzureSdk and GraphApi logs from Pulumi dynamic components to a null handler to avoid having them appear in the Pulumi output.

### :closed_umbrella: Related issues

Closes #1971 

### :microscope: Tests

Tested on a fresh SRE

### CODE REVIEW

File: data_safe_haven/logging/logger.py [hunk 1]
```diff
@@ -20,7 +21,9 @@ def get_console_handler() -> RichHandler:
     return next(h for h in get_logger().handlers if isinstance(h, RichHandler))
 
 
-def get_logger() -> logging.Logger:
+def get_logger(*, disable_logging: bool = False) -> logging.Logger:
+    if disable_logging:
+        return NonLoggingSingleton()
     return logging.getLogger("data_safe_haven")
```
Comment (JimMadge) [inline data_safe_haven/logging/logger.py]: I wonder if

```suggestion
def get_logger() -> logging.Logger:
    return logging.getLogger("data_safe_haven")
    
    
def get_null_logger() -> logging.Logger:
    return NonLoggingSingleton()
```

Would be clearer. Or changing the argument from `disable_logging` to `log_to_null`?
Saying, give me a logger and disable logging just sounds confusing to me.

Comment (JimMadge) [reply data_safe_haven/logging/logger.py]: Having things like `sdk = AzureSDK(disable_logging=True)` is good though.

File: data_safe_haven/logging/non_logging_singleton.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+class NonLoggingSingleton(logging.Logger, metaclass=Singleton):
+    """
+    Non-logging singleton that can be used by anything needing logs to be consumed
+    """
+
+    def __init__(self) -> None:
+        super().__init__(name="non-logger", level=logging.CRITICAL + 10)
+        while self.handlers:
+            self.removeHandler(self.handlers[0])
```
Comment (JimMadge) [inline data_safe_haven/logging/non_logging_singleton.py:14]: Could we add some tests for this.

Things like,

- Check the logging level of the object
- Check that it is a singleton?
- Ensuring that `logger.info()` _etc._ doesn't produce and file or terminal output.

**Review by JimMadge — CHANGES_REQUESTED** [review]
Looks good, just a couple of suggestions.

File: data_safe_haven/logging/logger.py (hunk 1 above)
Comment (jemrobinson) [reply data_safe_haven/logging/logger.py]: Done in 1aa489063

File: data_safe_haven/logging/non_logging_singleton.py (hunk 2 above)
Comment (jemrobinson) [reply data_safe_haven/logging/non_logging_singleton.py:14]: Done in d90adf6

**Review by JimMadge — APPROVED** [review]

**Pull request merged**